### PR TITLE
Spark Data Loader Function Abstraction and JDBC support

### PR DESCRIPTION
**Pull request by Yuqing-cat**
- Try to abstract all adhoc spark read if else code into one centralized place. 
- Add FileFormat and SQLType to help identify data loader functions to use, may not be the best practise.  
- Questions for existing code:
 1. What is the current logic of Local test and Path List. (cannot easily abstract path list related functions, so mark it as a Todo)
 2. Is there any specific reason that we don't leverage data loader in those

### CODE REVIEW

File: src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala [hunk 1]
```diff
@@ -637,35 +653,44 @@ private[offline] object SourceUtils {
...
+        if (ss.sparkContext.isLocal){
+          getTestDF(ss, inputData.inputPath)
+        } else {
+          loadAsDataFrame(ss, inputData.inputPath)
+        }
       }
     }
   }
 
+  // def loadSingleFileDF(ss: SparkSession, )
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala:688]: Delete

**Review by xiaoyongzhu — DISMISSED** [review]

File: src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala [hunk 2]
```diff
@@ -278,18 +280,32 @@ private[offline] object SourceUtils {
...
+    }
+  }
+
+  /**
+   * A Common Function to load Test DFs
+   * Will be moved to Test Utils later
+   * @param ss    Spark Session
+   * @param path  File Path
+   */
+  def getTestDF(ss: SparkSession, path: String): DataFrame = {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala:297]: nit: getTestDF -> getLocalDF

File: src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala [hunk 3]
```diff
@@ -637,35 +653,42 @@ private[offline] object SourceUtils {
...
           loadAsUnionDataFrame(ss, pathList)
-        } catch {
-          case _: Throwable => loadSeparateJsonFileAsAvroToDF(ss, inputData.inputPath).get
         }
-      } else {
-        loadAsUnionDataFrame(ss, pathList)
+      }
+      case FileFormat.JDBC => {
+        JDBCUtils.loadDataFrame(ss, inputData.inputPath)
+      }
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/util/SourceUtils.scala:674]: how do you load JDBC test data from local?

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SqlServerDataLoader.scala [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.linkedin.feathr.offline.source.dataloader.jdbc
+
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+class SqlServerDataLoader(ss: SparkSession) extends JDBCConnector(ss){
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SqlServerDataLoader.scala:5]: class doc

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SqlServerDataLoader.scala [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+  override val format = "com.microsoft.sqlserver.jdbc.spark"
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SqlServerDataLoader.scala:6]: doc

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package com.linkedin.feathr.offline.source.dataloader.jdbc
+
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+sealed trait SQLType
+
+object SQLType {
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala:7]: class doc

Comment (hangfei) [reply src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala:7]: the naming is confusing with Spark SQL Type. Make it more specific, like SqlDatabaseType

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  case object SqlServer extends SQLType
+  case object MySql extends SQLType
+  case object Postgres extends SQLType
+  case object DefaultJDBC extends SQLType
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala:11]: add doc for each of them.

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+  def getType (url: String): SQLType = url match {
+    case url if url.startsWith("jdbc:sqlserver") => SqlServer
+    case url if url.startsWith("jdbc:mysql") => MySql
+    case url if url.startsWith("jdbc:postgresql:") => Postgres
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/SQLType.scala:16]: add doc to explain why for future devs

File: src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package com.linkedin.feathr.offline.generation
+
+import com.linkedin.feathr.offline.source.dataloader.hdfs.FileFormat
+import org.apache.avro.generic.GenericRecord
+import org.apache.hadoop.mapred.JobConf
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, SaveMode}
+
+object SparkIOUtils {
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/SparkIOUtils.scala:9]: SparkIoUtils

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/hdfs/FileFormat.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import com.linkedin.feathr.offline.source.dataloader._
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.DataFrame
+
+
+object FileFormat {
+  private lazy val ss = SparkSession.builder.getOrCreate
+
+  val CSV = "CSV"
+  val AVRO_JSON = "AVRO_JSON"
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/hdfs/FileFormat.scala:13]: doc this

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/hdfs/FileFormat.scala [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  val AVRO = "AVRO"
+  val ORC = "ORC"
+  val PARQUET = "PARQUET"
+  val PATHLIST = "PATHLIST"
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/hdfs/FileFormat.scala:17]: doc this

**Review by hangfei — COMMENTED** [review]
how the users use this to get JDBC data in the API sidde?

File: build.sbt [hunk 12]
```diff
@@ -41,6 +41,13 @@ val localAndCloudCommonDependencies = Seq(
 
 ) // Common deps
 
+val jdbcDrivers = Seq(
+  "com.microsoft.sqlserver" % "mssql-jdbc" % "10.2.0.jre8",
```
[the remark this answers is not in the document]
Comment (Yuqing-cat) [reply build.sbt:45]: for 11, it needs higher version according to this doc: https://docs.microsoft.com/en-us/sql/connect/jdbc/using-the-jdbc-driver?view=sql-server-ver15

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/JDBCConnector.scala [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  /**
+   * jdbc Options needs to contain necessary information except jdbc url, include:
+   * driver option, authentication options, dbTable option, etc.
+   * extend a new class when new JDBC data connector is added.
+   * @param jdbcOptions
+   * @return
+   */
+  def getDFReader(jdbcOptions: Map[String, String]): DataFrameReader = {
+    _ss.read.format(format).options(jdbcOptions)
+  }
```
[the remark this answers is not in the document]
Comment (Yuqing-cat) [reply src/main/scala/com/linkedin/feathr/offline/source/dataloader/jdbc/JDBCConnector.scala:18]: I want to split the idea of data connector out from specific dataframe loader.
The configuration and authentication should be reusable.

**Review by hangfei — COMMENTED** [review]
Do not modify the class name since it will make it difficult to merge it back to our codebase.

File: src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 14]
```diff
@@ -140,7 +140,7 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
 
     // create a data source from anchorAndDerivations/nullValueSource.avro.json
     val df = new AvroJsonDataLoader(ss, "nullValueSource.avro.json").loadDataFrame()
-    SparkIOUUtil.writeDataFrame(df, mockDataFolder + "/nullValueSource")
+    SparkIOUtils.writeDataFrame(df, mockDataFolder + "/nullValueSource")
```
[the remark this answers is not in the document]
Comment (Yuqing-cat) [reply src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:143]: Renaming from SparkIOUUtil to SparkIOUtils, I talked with Jay he said it's a typo, isn't it?

File: build.sbt (hunk 12 above)
[the remark this answers is not in the document]
Comment (hangfei) [reply build.sbt:45]: add comment to explain this.

**Review by hangfei — APPROVED** [review]
Approve the change with the promise that:
1. a client side config and e2e test will be done in next PR.
2. a way to do local testing with JDBC source.

**Review by xiaoyongzhu — APPROVED** [review]

**Pull request merged**